source/iterator/iterator.go

```diff
@@ -0,0 +1,254 @@
+	lastProcessedVal any
+	// table is a table name.
+	table string
+	// keyColumns is a name of the column that iterator will use for setting key in record.
+	keyColumns []string
+	// orderingColumn is a name of the column that iterator will use for sorting data.
+	orderingColumn string
+	// batchSize is a size of rows batch.
+	batchSize int
+	// schema is a schema name.
+	schema string
+	// columnTypes is a map with all column types.
+	columnTypes map[string]string
+}
+
+// Params is an incoming iterator params for the New function.
+type Params struct {
+	DB               *sqlx.DB
+	LastProcessedVal any
+	Table            string
+	KeyColumns       []string
+	OrderingColumn   string
+	BatchSize        int
+	Schema           string
+}
+
+// New creates a new instance of the iterator.
+func New(ctx context.Context, params Params) (*Iterator, error) {
+	iterator := &Iterator{
+		db:               params.DB,
+		lastProcessedVal: params.LastProcessedVal,
+		table:            params.Table,
+		keyColumns:       params.KeyColumns,
+		orderingColumn:   params.OrderingColumn,
+		batchSize:        params.BatchSize,
+		schema:           params.Schema,
+	}
+
+	err := iterator.populateKeyColumns(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("populate key columns: %w", err)
+	}
+
+	iterator.columnTypes, err = columntypes.GetColumnTypes(ctx, params.DB, params.Table, params.Schema)
+	if err != nil {
+		return nil, fmt.Errorf("get column types: %w", err)
+	}
+
+	err = iterator.loadRows(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("load rows: %w", err)
+	}
+
+	return iterator, nil
+}
+
+// HasNext returns a bool indicating whether the iterator has the next record to return or not.
+func (iter *Iterator) HasNext(ctx context.Context) (bool, error) {
+	return iter.hasNext(ctx)
+}
+
+// Next returns the next record.
+func (iter *Iterator) Next(_ context.Context) (sdk.Record, error) {
+	row := make(map[string]any)
+	if err := iter.rows.MapScan(row); err != nil {
+		return sdk.Record{}, fmt.Errorf("scan rows: %w", err)
+	}
+
+	transformedRow, err := columntypes.TransformRow(row, iter.columnTypes)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("transform row column types: %w", err)
+	}
+
+	if _, ok := transformedRow[iter.orderingColumn]; !ok {
+		return sdk.Record{}, fmt.Errorf("ordering column %q not found", iter.orderingColumn)
+	}
+
+	key := make(sdk.StructuredData)
+	for i := range iter.keyColumns {
+		val, ok := transformedRow[iter.keyColumns[i]]
+		if !ok {
+			return sdk.Record{}, fmt.Errorf("key column %q not found", iter.keyColumns[i])
+		}
+
+		key[iter.keyColumns[i]] = val
+	}
+
+	rowBytes, err := json.Marshal(transformedRow)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("marshal row: %w", err)
+	}
+
+	// set a new position into the variable,
+	// to avoid saving position into the struct until we marshal the position
+	positionBytes, err := json.Marshal(transformedRow[iter.orderingColumn])
```

Comment:
We've decided to split the iterator into the `snapshot` and `cdc` ones.
But I've decided to do not to use the field with the iterator name because the `latestSnapshotValue` position field is enough to determine which iterator is currently running.